### PR TITLE
Refactor the passing around of metrics and Config in the HIR lowering

### PR DESCRIPTION
**Pull request by ggevay**
Follow-up to #28688, see https://github.com/MaterializeInc/materialize/pull/28688#discussion_r1700432239

~The first 6 commits are in #28688.~ Edit: That one merged in the meantime, so rebased this on main.

### Motivation

   * This PR refactors existing code.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/sql/src/plan/lowering.rs [hunk 1]
```diff
@@ -733,18 +707,18 @@ impl HirRelationExpr {
                                     kind.clone(),
                                     oa,
                                     id_gen,
-                                    config.enable_new_outer_join_lowering,
-                                    config.enable_outer_join_null_filter,
+                                    context.config.enable_new_outer_join_lowering,
+                                    context.config.enable_outer_join_null_filter,
```
[the remark this answers is not in the document]
Comment (ggevay) [reply src/sql/src/plan/lowering.rs]: Done

File: src/sql/src/plan/lowering.rs [hunk 2]
```diff
@@ -184,8 +189,10 @@ impl HirRelationExpr {
                                 get_outer,
                                 &ColumnMap::empty(),
                                 &mut CteMap::new(),
-                                &config.into(),
-                                metrics,
+                                &Context {
+                                    config: &config.into(),
+                                    metrics,
+                                },
```
[the remark this answers is not in the document]
Comment (ggevay) [reply src/sql/src/plan/lowering.rs]: Done

File: src/sql/src/plan/lowering.rs [hunk 3]
```diff
@@ -150,6 +150,11 @@ impl From<&SystemVars> for Config {
     }
 }
 
+struct Context<'a> {
```
[the remark this answers is not in the document]
Comment (ggevay) [reply src/sql/src/plan/lowering.rs:154]: Done.

**Comment by ggevay** [comment]
Thanks for the review!

**Review by antiguru — APPROVED** [review]

**Pull request merged**